### PR TITLE
Add address methods

### PR DESCRIPTION
**Pull request by huwshimi**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7170/

**Comment by huwshimi** [comment]
QA: if you have a user set up...
- get your user and note the address and billing addresses you currently have: `app.payment.getUser('[YOUR USERNAME HERE]', (error, data) => {console.log(error, data)})`
- add an address `app.payment.addAddress('[YOUR USERNAME HERE]', {name: 'address1', line1: '1 Maple St', line2: 'Loonie Lane', state: 'Bunnyhug', city: 'Sasquatch', postcode: '90210', countryCode: 'CA'}, (error) => {console.log(error)})`
- add an billing address `app.payment.addBillingAddress('[YOUR USERNAME HERE]', {name: 'address2', line1: '1 Maple St', line2: 'Loonie Lane', state: 'Bunnyhug', city: 'Sasquatch', postcode: '90210', countryCode: 'CA'}, (error) => {console.log(error)})`
- get your user and note your new addresses: `app.payment.getUser('[YOUR USERNAME HERE]', (error, data) => {console.log(error, data)})`
- remove your address: app.payment.removeAddress('[YOUR USERNAME HERE]', 'address1', (error) => {console.log(error)})
- remove your billing address: app.payment.removeBillingAddress('[YOUR USERNAME HERE]', 'address2', (error) => {console.log(error)})
- get your user and note your new addresses are not there anymore: `app.payment.getUser('[YOUR USERNAME HERE]', (error, data) => {console.log(error, data)})`

File: jujugui/static/gui/src/app/jujulib/payment.js [hunk 1]
```diff
@@ -276,6 +276,94 @@ var module = module;
...
+        - countryCode {String} The address country code
+        - phones {Array} a list of phone number strings
+      @param callback {Function} A callback to handle errors. Must accept an
+        error message or null as its first parameter.
+    */
+    addAddress: function(username, address, callback) {
+      const handler = error => {
+        callback(error);
+      };
+      const url = `${this.url}/u/${username}/addresses`;
```
Comment (jaycee) [inline jujugui/static/gui/src/app/jujulib/payment.js]: I did not know this sort of string interpolation was available in js. Neat.

File: jujugui/static/gui/src/app/jujulib/test-payment.js [hunk 2]
```diff
@@ -586,4 +591,238 @@ describe('jujulib payment service', function() {
...
+        city: 'Sasquatch',
+        postcode: '90210',
+        countryCode: 'CA'
+      };
+    });
+
+    it('can add an address', () => {
+      const originalMakeRequest = jujulib._makeRequest;
+      const makeRequest = sinon.stub();
+      jujulib._makeRequest = makeRequest;
```
Comment (jaycee) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: I believe you can do this with:
```
const makeRequest = sinon.stub(jujulib, '_makeRequest');
```
And then clean it up later with `makeRequest.restore()`.

It looks like you've used this pattern in most of your tests, so you could redo them all like this.

**Comment by jaycee** [comment]
QA OK. And :100: for qa directions that allow copy paste. :smile:

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7181/

**Comment by huwshimi** [comment]
Thanks @jaycee :shipit: to keep moving.

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**